File: scala/Board.scala
// $Id$
// Copyright 2008 John Hurst
// John Hurst (jbhurst@attglobal.net)
// 2008-12-24

class Board(s: Int) {
  require(s >= 4)
  val size = s
  private val pos = new Array[Int](size)
  private val col = new Array[Boolean](size)
  private val diag1 = new Array[Boolean](2 * size + 1)
  private val diag2 = new Array[Boolean](2 * size + 1)
  for (i <- 0 until size) pos(i) = -1

  def place(row: Int, col: Int) = {
    pos(row) = col
    this.col(col) = true
    diag1(col - row + size) = true
    diag2(col + row) = true
  }

  def unplace(row: Int) = {
    val col = pos(row)
    pos(row) = -1
    this.col(col) = false
    diag1(col - row + size) = false
    diag2(col + row) = false
    col
  }

  def isOk(row: Int, col: Int) =
    !(this.col(col) || diag1(col - row + size) || diag2(col + row))

  def solve = {
    var row = 0
    var col = 0
    while ((0 until size).contains(row)) {
      while (col < size && !isOk(row, col)) {
        col += 1
      }
      if (col < size) {
        place(row, col)
        row += 1
        col = 0
      }
      else {
        row -= 1
        if (row >= 0) {
          col = unplace(row) + 1
        }
      }
    }
    row == size
  }

  override def toString = (
    for (row <- 0 until size)
      yield " " * pos(row) + "*"
    ).mkString("\n")
}

File: ruby_queens/board.rb
# $Id: board.rb 664 2007-03-28 08:00:36Z jhurst $
# Copyright 2007 John Hurst
# John Hurst (jbhurst@attglobal.net)
# 2007-03-27 

class Board

  def initialize(size)
    @pos = Array.new(size, nil)
    @col = Array.new(size, false)
    @diag1 = Array.new(2*size, false)
    @diag2 = Array.new(2*size, false)
  end

  def size
    @pos.size
  end

  def col(row)
    @pos[row]
  end

  def place!(row, col)
    @pos[row] = col
    @col[col] = true
    @diag1[col-row+size] = true
    @diag2[col+row] = true
  end

  def unplace!(row)
    col = @pos[row]
    @pos[row] = nil
    @col[col] = false
    @diag1[col-row+size] = false
    @diag2[col+row] = false
    col
  end

  def ok?(row, col)
    return false if @col[col]
    return false if @diag1[col-row+size]
    return false if @diag2[col+row]
    true
  end

  def solve!
    row = 0
    col = 0
    while row >= 0 and row < size
      while col < size and !ok?(row, col)
        col += 1
      end
      if col < size
        place!(row, col)
        row += 1
        col = 0
      else
        row -= 1
        if row >= 0
          col = unplace!(row) + 1
        end
      end
    end
    return row == size
  end

  def printout
    @pos.each {|p| puts " " * p + "X"}
  end

end # class Board

